libs/contrib/Control/ST/File.idr

```diff
@@ -0,0 +1,215 @@
+module Control.ST.File 
+
+import Control.ST
+import Control.IOExcept
+
+%default total
+
+||| A record of the file modes that can read from a file.
+public export
+data ValidModeRead : Mode -> Type where
+  VMRRead   : ValidModeRead Read
+  VMRReadW  : ValidModeRead ReadWrite
+  VMRReadWT : ValidModeRead ReadWriteTruncate
+  VMRReadA  : ValidModeRead ReadAppend
+
+||| A record of the file modes that can write from a file.
+public export
+data ValidModeWrite : Mode -> Type where
+  VMWWrite  : ValidModeWrite WriteTruncate
+  VMWAppend : ValidModeWrite Append
+  VMWReadW  : ValidModeWrite ReadWrite
+  VMWReadWT : ValidModeWrite ReadWriteTruncate
+
+public export
+data FileHandle : (mode : Mode) -> Type where
+  FH : File -> FileHandle mode
+
+public export
```

Comment:
Perhaps it could be nice with a comment about the complete interface as well.

Reply:
Will also add the following description to File interface 
```Idris
||| Description of state transition operations on a file.
||| 
||| +------------+-------------+----------------+----------------+----------------+
||| | Operation  | State in    | Mode in        | State out      | Mode out       |
||| +------------+-------------+----------------+----------------+----------------+
||| | open       | -           | -              | file handle    | input paramater| 
||| | close      | file handle | any            | -              | -              | 
||| | eof        | file handle | ValidModeRead  | file handle    | ValidModeRead  | 
||| | flush      | file handle | any            | file handle    | unchanged      | 
||| | readLine   | file handle | ValidModeRead  | file handle    | ValidModeRead  | 
||| | readChar   | file handle | ValidModeRead  | file handle    | ValidModeRead  | 
||| | readFile   | -           | -              | -              | -              | 
||| | writeString| file handle | ValidModeWrite | file handle    | ValidModeWrite | 
||| | writeLine  | file handle | ValidModeWrite | file handle    | ValidModeWrite | 
||| | writeFile  | -           | -              | -              | -              | 
||| +------------+-------------+----------------+----------------+----------------+
||| 
```

Reply:
That sounds good